### PR TITLE
Sorting: Include previous priority

### PR DESCRIPTION
**Pull request by danielbeardsley**
It turns out there are some ugly edge cases with regard to choosing
which issues from one bucket to move into another when balancing the
buckets.

Specifically, If a p5 issue gets bumped to p4 because of rebalancing,
it essentially ends up at the top of the p4 bucket and will be the
first to jump from p4 -> p3 on further rebalancing. This is because we
sort issues by bucket, then priority label date. Because of this, two
bad things happen:

* Issues near the top of their bucket can easily be promoted all the
way up the chain to the highest bucket after a few rounds of balancing
* Issues not near the top of their buckets will never move up because
all issues coming in (regardless of from which bucket) will be on top
of them.

Solution
===========

@davidrans @sctice @danielj41 and @danielbeardsley all talked about
methods of dealing with this and came up with only one solution that
seemed better than the current one:

Sort issues by:

* Bucket priority, ASC
* Then *previous bucket trend*, DESC
   * Previous is the bucket the issue *used* to be in if it has
     ever moved, defaulting to the current bucket.
   * The trend is: came from higher priority (-1), has always been
     this priority (0), came from lower priority (1)
* Then issue creation date, _ascending_ (older issues are more
  important or at least should be looked at and intentionally demoted)

This way, any time an issue jumps the border from one bucket to the
next, it will be the first issue to jump back across the border.

**This makes the sort stable even after multiple rebalancings.**
No development branch specified; set it with:
git config feature.development-branch "name-of-branch"

Closes ifixit/ifixit#18503

### CODE REVIEW

**Comment by danielbeardsley** [comment]
CR should be done one commit at time (for more fun-like happy good time experience)

File: issue.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-var priorityLabelRegex = /^p([1-9])$/;
+var priorityLabelRegex = /^p([0-9]+)$/;
```
Comment (soriat) [inline issue.js]: This change also means that `p0`s can be rebalanced now right? Is that something we want?

File: github.js [hunk 2]
```diff
@@ -29,10 +29,43 @@ exports.addLabel = function addLabel(issue, label) {
...
+ */
+function getAllPages(currentPage, allResults) {
+   debug("Got %s results", currentPage.length);
+   return new Promise(function (resolve, reject) {
+      allResults = allResults || [];
+      allResults.push(currentPage);
+
+      if (!github.hasNextPage(currentPage)) {
+         debug("No more results");
+         return resolve([].concat(...allResults));
```
Comment (soriat) [inline github.js:75]: nice

File: issue.js (hunk 1 above)
Comment (danielbeardsley) [reply issue.js]: Oooh, good point. No.

Comment (danielbeardsley) [reply issue.js]: Fixed!

File: issue.js [hunk 3]
```diff
@@ -12,6 +14,15 @@ function Issue(number) {
       return priorityLabel && getNumber(priorityLabel);
    };
 
+   this.getPriorityLabel = function() {
+      return priorityLabel && priorityLabel.title;
+   };
+
+   this.getPreviousPriority = function() {
+      return previousPriorityLabel && getNumber(previousPriorityLabel)
+       || this.getPriority();;
```
Comment (soriat) [inline issue.js]: This expression is kinda hard to follow. This is basically saying that if `previousPriorityLabel` is truthy, then return `getNumber(previousPriorityLabel)`, otherwise return `this.getPriority()` right? Maybe some parens would make this clearer.

Also there's a double semicolon.

File: github.js [hunk 4]
```diff
@@ -40,6 +41,15 @@ exports.getOpenBacklogIssues = function() {
    }).then(getAllPages);
 };
 
+exports.getIssueEvents = function(issue) {
+   debug("Getting events for issue %s", issue.getNumber());
+   return getIssueEvents({
```
Comment (soriat) [inline github.js]: Is there a reason we're not using `includeDefaultParams` here?

File: load-issues.js [hunk 5]
```diff
@@ -2,20 +2,58 @@ var Issue = require('./issue.js');
 var config = require('./config.js');
 var debug = require('debug')('backlog:load-issues');
 var github = require('./github.js');
+var paraflow = require('paraflow');
 
-module.exports = function getIssues() {
+// Max number of github requests in parallel
+const PARALELL_LIMIT = 5;
```
Comment (soriat) [inline load-issues.js]: `PARALELL_LIMIT` => `PARALLEL_LIMIT`

File: load-issues.js [hunk 6]
```diff
@@ -2,20 +2,58 @@ var Issue = require('./issue.js');
...
+
+module.exports = () => {
    return github.getOpenBacklogIssues()
    .then(createIssuesFromGithubResponses)
+   .then(getPreviousLabelsForIssues)
 }
 
 function createIssuesFromGithubResponses(githubIssues) {
-   return githubIssues.map(createIssueFromGithubResponse);
+   return githubIssues.map(createIssueFromGithubResponse)
```
Comment (soriat) [inline load-issues.js]: Why did you remove the semicolons in this file?

File: test/issue-test.js [hunk 7]
```diff
@@ -27,6 +27,32 @@ describe('Issue', function() {
...
+         i.addLabel('p2', 2);
+         return i;
+      }
+      it('should choose the most recent previous label', function() {
+         var i = issueWithLabels();
+         assert.equal(4, i.getPreviousPriority());
+         assert.equal(9, i.getPriority());
+      });
+      it('should return null if there is no priority at all', function() {
+         var i = new Issue(1);
```
Comment (soriat) [inline test/issue-test.js:46]: Does it matter that issue #1 doesn't exist on the ifixit/ifixit repo?

File: issue.js (hunk 3 above)
Comment (danielbeardsley) [reply issue.js]: Yes, that's correct. Sure. parens for clarity are fine.

File: github.js (hunk 4 above)
Comment (danielbeardsley) [reply github.js]: Yeah, the issue number param is `issue_number` instead of just `number` like all the others. :-(

File: test/sort-issues-test.js [hunk 8]
```diff
@@ -20,17 +20,18 @@ describe('sort-issues', function() {
       assert.deepEqual(sorted, [i1,i2,i3]);
    });
 
-   it('should sort issues according to ASC priority, then appliedOn DESC', function() {
-      var i1 = issue('p1', 3);
+   it('should sort issues by createdOn ASC', function() {
+      var i1 = issue('p1', 1);
       var i2 = issue('p1', 2);
-      var i3 = issue('p1', 1);
+      var i3 = issue('p1', 3);
```
Comment (soriat) [inline test/sort-issues-test.js:26]: Might as well put an i4 in there that has the same `createdOn` as another one.

**Comment by soriat** [comment]
Left a couple comments that are mostly just clarification questions. Looks great! CR ⚡️

File: load-issues.js (hunk 5 above)
Comment (danielbeardsley) [reply load-issues.js]: Fixed

File: load-issues.js (hunk 6 above)
Comment (danielbeardsley) [reply load-issues.js]: Cause I'm trying to get used to not seeing them. They aren't needed in 99% of cases and I'm trying to let go of my FUD about seeing them :-). No real functional reason.

File: test/issue-test.js (hunk 7 above)
Comment (danielbeardsley) [reply test/issue-test.js:46]: No, there's no API requests here, this is just a unit test.

File: test/sort-issues-test.js (hunk 8 above)
Comment (danielbeardsley) [reply test/sort-issues-test.js:26]: I cover that case (same p-label and createdOn) in the test above.

Comment (danielbeardsley) [reply test/sort-issues-test.js:26]: Oh, and in the case of same p, previous, and createdOn, the sort behavior is undefined, so there's not much to test.

**Comment by soriat** [comment]
CR ⚡️

File: load-issues.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
+var Issue = require('./issue.js');
+var config = require('./config.js');
+var debug = require('debug')('backlog:load-issues');
+var github = require('./github.js');
+
+module.exports = function getIssues() {
```
Comment (sctice) [inline load-issues.js]: FYI, `x = function f() { … }` doesn't define the name `f`, even in an uninitialized state; it just defines `x` and assigns an anonymous function to it.

```js
module.exports = function f(x) {
  console.log(x)
}

console.log(f) // Error
```

File: load-issues.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+function createIssuesFromGithubResponses(githubIssues) {
+   return githubIssues.map(createIssueFromGithubResponse);
+}
+
+function createIssueFromGithubResponse(ghIssue) {
+   var issue = new Issue(ghIssue.number);
+   ghIssue.labels.forEach((label) => {
+      issue.addLabel(label.name, 1);
```
Comment (sctice) [inline load-issues.js]: The literal `1` here is totally opaque to me.

**Comment by sctice** [comment]
@soriat didn't leave anything to comment on! Looks good. CR 🎏

File: load-issues.js (hunk 9 above)
Comment (danielbeardsley) [reply load-issues.js]: Kinda, it actually defines it within the function itself so you can do recursion. It also shows up in stack traces instead of "anonymous function".  In this case, I did it for the sake of the human reader.

https://stackoverflow.com/a/3883869/13216

File: load-issues.js (hunk 10 above)
Comment (danielbeardsley) [reply load-issues.js]: Good point. Fixed with a comment. I went down the road of removing the `appliedOn` argument to `addLabel()` since it's not used in sorting. It was a lot of little changes and the code wasn't much simpler. Filling in the correct data from the issue events wouldn't be that hard though.

**Comment by soriat** [comment]
CR ⚡️

**Comment by danielbeardsley** [comment]
Given that I just updated a comment to improve the clarity where @sctice said it wasn't clear, I'm gonna say this is still CR 👍

**Comment by danielbeardsley** [comment]
From a dry-run: https://www.cominor.com/Misc/pastebin/523

```
Tue, 21 Nov 2017 17:37:22 GMT backlog:github Getting all open backlog issues
Tue, 21 Nov 2017 17:37:23 GMT backlog:github Got 100 results
Tue, 21 Nov 2017 17:37:24 GMT backlog:github Got 100 results
Tue, 21 Nov 2017 17:37:25 GMT backlog:github Got 55 results
Tue, 21 Nov 2017 17:37:25 GMT backlog:github No more results
Tue, 21 Nov 2017 17:37:25 GMT backlog:load-issues Retrieving label events for 255 issues
Tue, 21 Nov 2017 17:37:25 GMT backlog:github Getting events for issue 21971
Tue, 21 Nov 2017 17:37:25 GMT backlog:github Getting events for issue 21970
Tue, 21 Nov 2017 17:37:25 GMT backlog:github Getting events for issue 21964
...
Tue, 21 Nov 2017 17:37:26 GMT backlog:github Getting events for issue 21866
Tue, 21 Nov 2017 17:37:26 GMT backlog:github Getting events for issue 21861
Tue, 21 Nov 2017 17:37:26 GMT backlog:issue Issue 21929 found previous priority label: p3
Tue, 21 Nov 2017 17:37:26 GMT backlog:github Getting events for issue 21843
Tue, 21 Nov 2017 17:37:26 GMT backlog:github Getting events for issue 21839
Tue, 21 Nov 2017 17:37:26 GMT backlog:github Getting events for issue 21836
Tue, 21 Nov 2017 17:37:26 GMT backlog:issue Issue 21866 found previous priority label: p2
Tue, 21 Nov 2017 17:37:26 GMT backlog:github Getting events for issue 21831
Tue, 21 Nov 2017 17:37:26 GMT backlog:issue Issue 21861 found previous priority label: p3
Tue, 21 Nov 2017 17:37:26 GMT backlog:github Getting events for issue 21828
Tue, 21 Nov 2017 17:37:26 GMT backlog:github Getting events for issue 21820
Tue, 21 Nov 2017 17:37:26 GMT backlog:issue Issue 21831 found previous priority label: p3
Tue, 21 Nov 2017 17:37:26 GMT backlog:github Getting events for issue 21812
Tue, 21 Nov 2017 17:37:26 GMT backlog:github Getting events for issue 21808
Tue, 21 Nov 2017 17:37:26 GMT backlog:github Getting events for issue 21800
Tue, 21 Nov 2017 17:37:26 GMT backlog:issue Issue 21808 found previous priority label: p3
Tue, 21 Nov 2017 17:37:26 GMT backlog:github Getting events for issue 21777
Tue, 21 Nov 2017 17:37:26 GMT backlog:github Getting events for issue 21776
Tue, 21 Nov 2017 17:37:26 GMT backlog:issue Issue 21800 found previous priority label: p3
Tue, 21 Nov 2017 17:37:26 GMT backlog:github Getting events for issue 21767
Tue, 21 Nov 2017 17:37:26 GMT backlog:issue Issue 21839 found previous priority label: p3
Tue, 21 Nov 2017 17:37:26 GMT backlog:github Getting events for issue 21747
Tue, 21 Nov 2017 17:37:26 GMT backlog:github Getting events for issue 21746
Tue, 21 Nov 2017 17:37:26 GMT backlog:github Getting events for issue 21745
Tue, 21 Nov 2017 17:37:26 GMT backlog:issue Issue 21812 found previous priority label: p2
Tue, 21 Nov 2017 17:37:26 GMT backlog:github Getting events for issue 21744
Tue, 21 Nov 2017 17:37:27 GMT backlog:issue Issue 21767 found previous priority label: p3
Tue, 21 Nov 2017 17:37:27 GMT backlog:github Getting events for issue 21743
Tue, 21 Nov 2017 17:37:27 GMT backlog:issue Issue 21747 found previous priority label: p3
Tue, 21 Nov 2017 17:37:27 GMT backlog:github Getting events for issue 21742
Tue, 21 Nov 2017 17:37:27 GMT backlog:github Getting events for issue 21739
Tue, 21 Nov 2017 17:37:27 GMT backlog:github Getting events for issue 21724
Tue, 21 Nov 2017 17:37:27 GMT backlog:issue Issue 21744 found previous priority label: p3
...
Tue, 21 Nov 2017 17:37:39 GMT backlog:load-issues Done processing label events for 255 issues
Tue, 21 Nov 2017 17:37:39 GMT backlog:filter Filtering 255 issues; removing those without priority
Tue, 21 Nov 2017 17:37:39 GMT backlog:filter Removed 15 issues without priorities
Tue, 21 Nov 2017 17:37:39 GMT backlog:sort Sorting 240 issues by priority, then previous priority, then creation date
Tue, 21 Nov 2017 17:37:39 GMT backlog:sort Priority comparisons: 751
Tue, 21 Nov 2017 17:37:39 GMT backlog:sort Previous Priority comparisons: 363
Tue, 21 Nov 2017 17:37:39 GMT backlog:sort Created On comparisons: 227
Tue, 21 Nov 2017 17:37:39 GMT backlog:bucketize Putting 240 issues into 5 buckets
Tue, 21 Nov 2017 17:37:39 GMT backlog:bucketize Bucket sizes: [48,48,48,48,48]
Tue, 21 Nov 2017 17:37:39 GMT backlog:github [dry run] Issue 19259: removing label p4
Tue, 21 Nov 2017 17:37:39 GMT backlog:github [dry run] Issue 19259: adding label p5
Tue, 21 Nov 2017 17:37:39 GMT backlog:github [dry run] Issue 17626: removing label p4
Tue, 21 Nov 2017 17:37:39 GMT backlog:github [dry run] Issue 17626: adding label p5
Tue, 21 Nov 2017 17:37:39 GMT backlog:github [dry run] Issue 18207: removing label p3
Tue, 21 Nov 2017 17:37:39 GMT backlog:github [dry run] Issue 18207: adding label p4
Tue, 21 Nov 2017 17:37:39 GMT backlog:github [dry run] Issue 17508: removing label p3
Tue, 21 Nov 2017 17:37:39 GMT backlog:github [dry run] Issue 17508: adding label p4
Tue, 21 Nov 2017 17:37:39 GMT backlog:github [dry run] Issue 20518: removing label p3
Tue, 21 Nov 2017 17:37:39 GMT backlog:github [dry run] Issue 20518: adding label p4
Tue, 21 Nov 2017 17:37:39 GMT backlog:github [dry run] Issue 21800: removing label p2
Tue, 21 Nov 2017 17:37:39 GMT backlog:github [dry run] Issue 21800: adding label p3
Tue, 21 Nov 2017 17:37:39 GMT backlog:github [dry run] Issue 15870: removing label p2
Tue, 21 Nov 2017 17:37:39 GMT backlog:github [dry run] Issue 15870: adding label p3
Tue, 21 Nov 2017 17:37:39 GMT backlog:github [dry run] Issue 21929: removing label p2
Tue, 21 Nov 2017 17:37:39 GMT backlog:github [dry run] Issue 21929: adding label p3
Tue, 21 Nov 2017 17:37:39 GMT backlog:github [dry run] Issue 16825: removing label p2
Tue, 21 Nov 2017 17:37:39 GMT backlog:github [dry run] Issue 16825: adding label p3
Tue, 21 Nov 2017 17:37:39 GMT backlog:app Balancing complete
```

With the important parts being:
```
Tue, 21 Nov 2017 17:37:39 GMT backlog:load-issues Done processing label events for 255 issues
Tue, 21 Nov 2017 17:37:39 GMT backlog:filter Filtering 255 issues; removing those without priority
Tue, 21 Nov 2017 17:37:39 GMT backlog:filter Removed 15 issues without priorities
Tue, 21 Nov 2017 17:37:39 GMT backlog:sort Sorting 240 issues by priority, then previous priority, then creation date
Tue, 21 Nov 2017 17:37:39 GMT backlog:sort Priority comparisons: 751
Tue, 21 Nov 2017 17:37:39 GMT backlog:sort Previous Priority comparisons: 363
Tue, 21 Nov 2017 17:37:39 GMT backlog:sort Created On comparisons: 227
Tue, 21 Nov 2017 17:37:39 GMT backlog:bucketize Putting 240 issues into 5 buckets
Tue, 21 Nov 2017 17:37:39 GMT backlog:bucketize Bucket sizes: [48,48,48,48,48]
```

I think this is QA 👍
